File: clients/go/internal/state/state.go
// Package state provides thread-safe state management for the Detrix client.
package state

import (
	"fmt"
	"os"
	"os/exec"
	"sync"
)

// State represents the client state machine state.
type State string

const (
	// StateSleeping is the initial/inactive state.
	// Control plane running, debugger not started, not registered with daemon.
	StateSleeping State = "sleeping"

	// StateWaking is the transitional state during wake operation.
	// Debugger starting, daemon registration in progress.
	StateWaking State = "waking"

	// StateAwake is the active state.
	// Debugger listening, registered with daemon.
	StateAwake State = "awake"
)

// DelveProcess holds information about a running Delve process.
type DelveProcess struct {
	Cmd  *exec.Cmd
	Host string
	Port int
}

// ClientState holds all client state.
// Access must be synchronized using the embedded RWMutex.
type ClientState struct {
	sync.RWMutex

	// Configuration (set once in Init)
	Name        string
	ControlHost string
	ControlPort int
	DebugPort   int
	DaemonURL   string
	DelvePath   string
	DetrixHome  string
	SafeMode    bool // SafeMode: only logpoints allowed, no breakpoint operations

	// Timeouts (in milliseconds for simplicity)
	HealthCheckTimeoutMs int
	RegisterTimeoutMs    int
	UnregisterTimeoutMs  int
	DelveStartTimeoutMs  int

	// Runtime state
	State             State
	ActualControlPort int
	ActualDebugPort   int
	DebugPortActive   bool
	ConnectionID      *string

	// Process management
	DelveProcess *DelveProcess
}

// Global client state instance
var (
	globalState       *ClientState
	globalStateMu     sync.Mutex
	globalWakeLock    sync.Mutex
	globalInitialized bool
)

// Get returns the global client state, initializing if needed.
func Get() *ClientState {
	globalStateMu.Lock()
	defer globalStateMu.Unlock()

	if globalState == nil {
		globalState = &ClientState{
			State:                StateSleeping,
			HealthCheckTimeoutMs: 2000,
			RegisterTimeoutMs:    5000,
			UnregisterTimeoutMs:  2000,
			DelveStartTimeoutMs:  10000,
		}
	}
	return globalState
}

// Reset resets the global state to initial values.
func Reset() {
	globalStateMu.Lock()
	defer globalStateMu.Unlock()

	globalState = nil
	globalInitialized = false
}

// IsInitialized returns whether the client has been initialized.
func IsInitialized() bool {
	globalStateMu.Lock()
	defer globalStateMu.Unlock()
	return globalInitialized
}

// SetInitialized marks the client as initialized.
func SetInitialized(v bool) {
	globalStateMu.Lock()
	defer globalStateMu.Unlock()
	globalInitialized = v
}

// AcquireWakeLock acquires the wake lock to prevent concurrent wake operations.
func AcquireWakeLock() {
	globalWakeLock.Lock()
}

// ReleaseWakeLock releases the wake lock.
func ReleaseWakeLock() {
	globalWakeLock.Unlock()
}

// TryAcquireWakeLock tries to acquire the wake lock without blocking.
// Returns true if acquired, false if already held.
func TryAcquireWakeLock() bool {
	return globalWakeLock.TryLock()
}

// StatusSnapshot returns a snapshot of the current status.
// This is safe to call from any goroutine.
type StatusSnapshot struct {
	State           State
	Name            string
	ControlHost     string
	ControlPort     int
	DebugPort       int
	DebugPortActive bool
	DaemonURL       string
	ConnectionID    *string
}

// Snapshot returns a thread-safe snapshot of the current state.
func (s *ClientState) Snapshot() StatusSnapshot {
	s.RLock()
	defer s.RUnlock()

	var connID *string
	if s.ConnectionID != nil {
		v := *s.ConnectionID
		connID = &v
	}

	return StatusSnapshot{
		State:           s.State,
		Name:            s.Name,
		ControlHost:     s.ControlHost,
		ControlPort:     s.ActualControlPort,
		DebugPort:       s.ActualDebugPort,
		DebugPortActive: s.DebugPortActive,
		DaemonURL:       s.DaemonURL,
		ConnectionID:    connID,
	}
}

// GenerateConnectionName generates a connection name.
// If name is empty, generates "detrix-client-{pid}".
func GenerateConnectionName(name string) string {
	if name != "" {
		return name
	}
	return fmt.Sprintf("detrix-client-%d", os.Getpid())
}

File: clients/rust/src/state.rs
//! Thread-safe state management for the Detrix client.

use std::sync::{Mutex, MutexGuard, OnceLock, RwLock};

use tracing::warn;

use crate::error::Error;
use crate::generated::{ClientState, StatusResponse};
use crate::lldb::LldbProcess;

/// Global client state singleton.
static GLOBAL_STATE: OnceLock<RwLock<InternalState>> = OnceLock::new();

/// Wake lock to prevent concurrent wake operations.
static WAKE_LOCK: OnceLock<Mutex<()>> = OnceLock::new();

/// Initialization flag.
static INITIALIZED: OnceLock<RwLock<bool>> = OnceLock::new();

/// Internal state of the Detrix client.
#[derive(Debug)]
pub struct InternalState {
    /// Current state machine state.
    pub state: ClientState,

    /// Connection name.
    pub name: String,

    /// Control plane host.
    pub control_host: String,

    /// Configured control port (0 = auto).
    pub control_port: u16,

    /// Actual control port after binding.
    pub actual_control_port: u16,

    /// Configured debug port (0 = auto).
    pub debug_port: u16,

    /// Actual debug port after lldb-dap starts.
    pub actual_debug_port: u16,

    /// Whether the debug port is currently active.
    pub debug_port_active: bool,

    /// Daemon URL.
    pub daemon_url: String,

    /// Connection ID from daemon registration.
    pub connection_id: Option<String>,

    /// Path to lldb-dap binary.
    pub lldb_dap_path: String,

    /// Detrix home directory.
    pub detrix_home: Option<String>,

    /// Safe mode enabled.
    pub safe_mode: bool,

    /// Health check timeout in milliseconds.
    pub health_check_timeout_ms: u64,

    /// Register timeout in milliseconds.
    pub register_timeout_ms: u64,

    /// Unregister timeout in milliseconds.
    pub unregister_timeout_ms: u64,

    /// lldb-dap start timeout in milliseconds.
    pub lldb_start_timeout_ms: u64,
}

impl Default for InternalState {
    fn default() -> Self {
        Self {
            state: ClientState::Sleeping,
            name: String::new(),
            control_host: "127.0.0.1".to_string(),
            control_port: 0,
            actual_control_port: 0,
            debug_port: 0,
            actual_debug_port: 0,
            debug_port_active: false,
            daemon_url: "http://127.0.0.1:8090".to_string(),
            connection_id: None,
            lldb_dap_path: String::new(),
            detrix_home: None,
            safe_mode: false,
            health_check_timeout_ms: 2000,
            register_timeout_ms: 5000,
            unregister_timeout_ms: 2000,
            lldb_start_timeout_ms: 10000,
        }
    }
}

impl InternalState {
    /// Create a StatusResponse from the current state.
    pub fn to_status_response(&self) -> StatusResponse {
        StatusResponse {
            state: self.state,
            name: self.name.clone(),
            control_host: self.control_host.clone(),
            control_port: i32::from(self.actual_control_port),
            debug_port: i32::from(self.actual_debug_port),
            debug_port_active: self.debug_port_active,
            daemon_url: self.daemon_url.clone(),
            connection_id: self.connection_id.clone(),
        }
    }
}

/// Get the global state, initializing if needed.
pub fn get() -> &'static RwLock<InternalState> {
    GLOBAL_STATE.get_or_init(|| RwLock::new(InternalState::default()))
}

/// Reset the global state to initial values.
pub fn reset() {
    if let Some(state) = GLOBAL_STATE.get() {
        match state.write() {
            Ok(mut guard) => *guard = InternalState::default(),
            Err(poisoned) => {
                warn!("Global state lock poisoned during reset, recovering");
                *poisoned.into_inner() = InternalState::default();
            }
        }
    }
    if let Some(init) = INITIALIZED.get() {
        match init.write() {
            Ok(mut guard) => *guard = false,
            Err(poisoned) => {
                warn!("Initialized lock poisoned during reset, recovering");
                *poisoned.into_inner() = false;
            }
        }
    }
}

/// Check if the client has been initialized.
pub fn is_initialized() -> bool {
    *INITIALIZED
        .get_or_init(|| RwLock::new(false))
        .read()
        .unwrap_or_else(|poisoned| {
            warn!("Initialized lock poisoned during read, recovering");
            poisoned.into_inner()
        })
}

/// Set the initialization flag.
pub fn set_initialized(value: bool) {
    let init = INITIALIZED.get_or_init(|| RwLock::new(false));
    match init.write() {
        Ok(mut guard) => *guard = value,
        Err(poisoned) => {
            warn!("Initialized lock poisoned during write, recovering");
            *poisoned.into_inner() = value;
        }
    }
}

/// Acquire the wake lock to prevent concurrent wake operations.
///
/// # Errors
///
/// Returns `Error::ControlPlaneError` if the lock is poisoned (a thread panicked
/// while holding it). Previously this would panic; now callers should handle
/// the error gracefully.
pub fn acquire_wake_lock() -> Result<MutexGuard<'static, ()>, Error> {
    Ok(WAKE_LOCK.get_or_init(|| Mutex::new(())).lock()?)
}

/// Global lldb-dap process handle.
static LLDB_PROCESS: OnceLock<Mutex<Option<LldbProcess>>> = OnceLock::new();

/// Get the lldb-dap process handle.
pub fn get_lldb_process() -> &'static Mutex<Option<LldbProcess>> {
    LLDB_PROCESS.get_or_init(|| Mutex::new(None))
}

/// Set the lldb-dap process.
pub fn set_lldb_process(process: LldbProcess) {
    match get_lldb_process().lock() {
        Ok(mut guard) => *guard = Some(process),
        Err(poisoned) => {
            warn!("lldb process lock poisoned during set, recovering");
            *poisoned.into_inner() = Some(process);
        }
    }
}

/// Take and clear the lldb-dap process.
pub fn take_lldb_process() -> Option<LldbProcess> {
    match get_lldb_process().lock() {
        Ok(mut guard) => guard.take(),
        Err(poisoned) => {
            warn!("lldb process lock poisoned during take, recovering");
            poisoned.into_inner().take()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_state() {
        let state = InternalState::default();
        assert!(matches!(state.state, ClientState::Sleeping));
        assert_eq!(state.control_host, "127.0.0.1");
    }

    #[test]
    fn test_status_response() {
        let mut state = InternalState::default();
        state.name = "test-client".to_string();
        state.actual_control_port = 8091;

        let response = state.to_status_response();
        assert_eq!(response.name, "test-client");
        assert_eq!(response.control_port, 8091);
    }

    #[test]
    fn test_set_initialized_roundtrip() {
        set_initialized(true);
        assert!(is_initialized());
        set_initialized(false);
        assert!(!is_initialized());
    }

    #[test]
    fn test_take_lldb_process_returns_none_when_empty() {
        // Reset any existing process
        let _ = take_lldb_process();
        assert!(take_lldb_process().is_none());
    }

    #[test]
    fn test_to_status_response_port_cast_safe() {
        let mut state = InternalState::default();
        // u16::MAX should safely convert to i32
        state.actual_control_port = u16::MAX;
        state.actual_debug_port = u16::MAX;

        let response = state.to_status_response();
        assert_eq!(response.control_port, i32::from(u16::MAX));
        assert_eq!(response.debug_port, i32::from(u16::MAX));
    }
}
